samples/subsys/usb/hid-cdc/src/main.c

```diff
@@ -0,0 +1,853 @@
+			void (*handler)(struct device*, struct gpio_callback*,
+			u32_t), struct gpio_callback *callback)
+{
+	if (!gpio) {
+		LOG_ERR("Could not find PORT");
+		return -ENXIO;
+	}
+	gpio_pin_configure(gpio, pin,
+			   GPIO_DIR_IN | GPIO_INT | GPIO_INT_DEBOUNCE |
+			   GPIO_INT_EDGE | flags);
+	gpio_init_callback(callback, handler, BIT(pin));
+	gpio_add_callback(gpio, callback);
+	gpio_pin_enable_callback(gpio, pin);
+	return 0;
+}
+
+void main(void)
+{
+	struct device *hid0_dev, *hid1_dev, *cdc0_dev, *cdc1_dev;
+	u32_t dtr = 0U;
+
+	/* Configure devices */
+
+	hid0_dev = device_get_binding(CONFIG_USB_HID_DEVICE_NAME_0);
+	if (hid0_dev == NULL) {
+		LOG_ERR("Cannot get USB HID 0 Device");
+		return;
+	}
+
+	hid1_dev = device_get_binding(CONFIG_USB_HID_DEVICE_NAME_1);
+	if (hid1_dev == NULL) {
+		LOG_ERR("Cannot get USB HID 1 Device");
+		return;
+	}
+
+	cdc0_dev = device_get_binding(CONFIG_CDC_ACM_PORT_NAME_0);
+	if (cdc0_dev == NULL) {
+		LOG_ERR("Cannot get USB CDC 0 Device");
+		return;
+	}
+
+	cdc1_dev = device_get_binding(CONFIG_CDC_ACM_PORT_NAME_1);
+	if (cdc1_dev == NULL) {
+		LOG_ERR("Cannot get USB CDC 1 Device");
+		return;
+	}
+
+	if (callbacks_configure(device_get_binding(PORT0), PIN0, PIN0_FLAGS,
+				&btn0, &callback[0])) {
+		LOG_ERR("Failed configuring button 0 callback.");
+		return;
+	}
+
+#ifdef SW1_GPIO_PIN
+	if (callbacks_configure(device_get_binding(PORT1), PIN1, PIN1_FLAGS,
+				&btn1, &callback[1])) {
+		LOG_ERR("Failed configuring button 1 callback.");
+		return;
+	}
+#endif
+
+#ifdef SW2_GPIO_PIN
+	if (callbacks_configure(device_get_binding(PORT2), PIN2, PIN2_FLAGS,
+				&btn2, &callback[2])) {
+		LOG_ERR("Failed configuring button 2 callback.");
+		return;
+	}
+#endif
+
+#ifdef SW3_GPIO_PIN
+	if (callbacks_configure(device_get_binding(PORT3), PIN3, PIN3_FLAGS,
+				&btn3, &callback[3])) {
+		LOG_ERR("Failed configuring button 3 callback.");
+		return;
+	}
+#endif
+
+	/* Initialize HID */
+
+	usb_hid_register_device(hid0_dev, hid_mouse_report_desc,
+				sizeof(hid_mouse_report_desc), &ops);
+
+	usb_hid_register_device(hid1_dev, hid_kbd_report_desc,
+				sizeof(hid_kbd_report_desc), &ops);
+	usb_hid_init(hid0_dev);
+	usb_hid_init(hid1_dev);
+
+	/* Initialize CDC ACM */
+
+	LOG_INF("Wait for DTR on CDC ACM 0");
+	while (1) {
+		uart_line_ctrl_get(cdc0_dev, LINE_CTRL_DTR, &dtr);
+		if (dtr) {
+			break;
+		}
+	}
+	LOG_INF("DTR on CDC ACM 0 set");
+
+	LOG_INF("Wait for DTR on CDC ACM 1");
+	while (1) {
+		uart_line_ctrl_get(cdc1_dev, LINE_CTRL_DTR, &dtr);
+		if (dtr) {
+			break;
+		}
+	}
+	LOG_INF("DTR on CDC ACM 1 set");
+
+	/* Wait 1 sec for the host to do all settings */
+	k_busy_wait(K_SECONDS(1));
+
+	uart_irq_callback_set(cdc0_dev, cdc_mouse_int_handler);
+	uart_irq_callback_set(cdc1_dev, cdc_kbd_int_handler);
+
+	write_data(cdc0_dev, banner0, strlen(banner0));
+	write_data(cdc1_dev, banner1, strlen(banner1));
+
+	uart_irq_rx_enable(cdc0_dev);
+	uart_irq_rx_enable(cdc1_dev);
+
+	struct app_evt_t *ev;
```

Comment:
We usually define variables at the beginning of the block, however this one is used in while() loop below, right? Move it here.

Reply:
Moved.